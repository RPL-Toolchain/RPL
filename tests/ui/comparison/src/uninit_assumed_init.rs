//@revisions: inline normal
//@[normal] compile-flags: -Zinline-mir=false
use std::mem::MaybeUninit;

#[cfg_attr(test, test)]
fn base_case() {
    #[expect(invalid_value)]
    let x: usize = unsafe { MaybeUninit::uninit().assume_init() };
    //~^ uninit_assumed_init
    dbg!(x);
}

#[cfg_attr(test, test)]
fn cross_function_uninit() {
    fn uninit<T>() -> MaybeUninit<T> {
        MaybeUninit::uninit()
    }
    let x: usize = unsafe { uninit().assume_init() };
    //~[inline]^ uninit_assumed_init
    dbg!(x);
}

#[cfg_attr(test, test)]
fn cross_function_assume_init() {
    unsafe fn assume_init<T>(maybe_uninit: MaybeUninit<T>) -> T {
        unsafe { maybe_uninit.assume_init() }
    }
    let x: usize = unsafe { assume_init(MaybeUninit::uninit()) };
    //~[inline]^ uninit_assumed_init
    dbg!(x);
}

#[cfg_attr(test, test)]
fn cross_statement() {
    let maybe_uninit = MaybeUninit::uninit();
    let x: usize = unsafe { maybe_uninit.assume_init() };
    //~^ uninit_assumed_init
    dbg!(x);
}

#[cfg_attr(test, test)]
pub(crate) fn main() {
    base_case();
    cross_function_uninit();
    cross_function_assume_init();
    cross_statement();
}
