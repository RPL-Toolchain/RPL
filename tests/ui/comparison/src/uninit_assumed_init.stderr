error: this call for this type may be undefined behavior
  --> tests/ui/comparison/src/uninit_assumed_init.rs:6:51
   |
LL |     let x: usize = unsafe { MaybeUninit::uninit().assume_init() };
   |                                                   ^^^^^^^^^^^^^
   |
   = note: `-D rpl::uninit-assumed-init` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::uninit_assumed_init)]`

error: this call for this type may be undefined behavior
  --> tests/ui/comparison/src/uninit_assumed_init.rs:16:38
   |
LL |     let x: usize = unsafe { uninit().assume_init() };
   |                                      ^^^^^^^^^^^^^

error: this call for this type may be undefined behavior
  --> tests/ui/comparison/src/uninit_assumed_init.rs:26:29
   |
LL |     let x: usize = unsafe { assume_init(MaybeUninit::uninit()) };
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: this call for this type may be undefined behavior
  --> tests/ui/comparison/src/uninit_assumed_init.rs:34:42
   |
LL |     let x: usize = unsafe { maybe_uninit.assume_init() };
   |                                          ^^^^^^^^^^^^^

error: aborting due to 4 previous errors

